server/core_notification.go

```diff
@@ -0,0 +1,215 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"bytes"
+	"database/sql"
+	"encoding/base64"
+	"encoding/gob"
+	"errors"
+	"strconv"
+	"strings"
+
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/heroiclabs/nakama/api"
+	"github.com/satori/go.uuid"
+	"go.uber.org/zap"
+)
+
+const (
+	NOTIFICATION_DM_REQUEST         int64 = -1
+	NOTIFICATION_FRIEND_REQUEST     int64 = -2
+	NOTIFICATION_FRIEND_ACCEPT      int64 = -3
+	NOTIFICATION_GROUP_ADD          int64 = -4
+	NOTIFICATION_GROUP_JOIN_REQUEST int64 = -5
+	NOTIFICATION_FRIEND_JOIN_GAME   int64 = -6
+)
+
+type notificationCacheableCursor struct {
+	NotificationID string
+}
+
+func NotificationSend(logger *zap.Logger, db *sql.DB, tracker Tracker, messageRouter MessageRouter, notifications map[uuid.UUID][]*api.Notification) error {
+	persistentNotifications := make(map[uuid.UUID][]*api.Notification)
+	for userID, n := range notifications {
+		for _, userNotification := range n {
+			// Select persistent notifications for storage.
+			if userNotification.Persistent {
+				pun := persistentNotifications[userID]
+				if pun == nil {
+					pun = make([]*api.Notification, 0)
+				}
+				pun = append(pun, userNotification)
+				persistentNotifications[userID] = pun
+			}
+		}
+	}
+
+	if len(persistentNotifications) > 0 {
+		if err := NotificationSave(logger, db, persistentNotifications); err != nil {
+			return err
+		}
+	}
+
+	// TODO (mo, zyro): Wire up Tracker to list presences by topic
+	//for userID, ns := range notifications {
+	//	nots := &rtapi.Notifications{Notifications: ns}
+	//	presences := tracker.ListByTopic("notifications:" + userID)
+	//	if len(presences) != 0 {
+	//		envelope := &rtapi.Envelope{Message: &rtapi.Envelope_Notifications{Notifications: nots}}
+	//		messageRouter.SendToPresences(logger, presences, envelope)
+	//	}
+	//}
+
+	return nil
+}
+
+func NotificationList(logger *zap.Logger, db *sql.DB, userID uuid.UUID, limit int64, cursor string) (*api.NotificationList, error) {
+	nc := &notificationCacheableCursor{}
+	if cursor != "" {
+		if cb, err := base64.RawURLEncoding.DecodeString(cursor); err != nil {
+			logger.Warn("Could not base64 decode notification cursor.", zap.String("cursor", cursor))
+			return nil, errors.New("Malformed cursor was used.")
+		} else {
+			if err := gob.NewDecoder(bytes.NewReader(cb)).Decode(nc); err != nil {
+				logger.Warn("Could not decode notification cursor.", zap.String("cursor", cursor))
+				return nil, errors.New("Malformed cursor was used.")
+			}
+		}
+	}
+
+	params := []interface{}{userID, limit}
+	cursorQuery := ""
+	if nc.NotificationID != "" {
+		cursorQuery = " AND id > $3::UUID "
+		params = append(params, nc.NotificationID)
+	}
+
+	rows, err := db.Query(`
+SELECT id, subject, content, code, sender_id, create_time
+FROM notification
+WHERE user_id = $1`+cursorQuery+`
```

Comment:
Notifications should be ordered by `create_time`, from newest to oldest I think? Previously expiry time achieved this implicitly, now we should probably order by `create_time` and make that part of the cursor too.

Reply:
I can do ORDER BY create_time ASC - but i don't think it is necessary to embed this in the cursor as the notification ID is deterministic and notification create_time timestamp is always an increasing value.